atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueServiceTest.java

```diff
@@ -0,0 +1,125 @@
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.Futures;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.Namespace;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.table.description.TableDefinition;
+import com.palantir.atlasdb.table.description.ValueType;
+import com.palantir.atlasdb.transaction.api.ConflictHandler;
+
+public class TableRemappingKeyValueServiceTest {
+    private static final Namespace NAMESPACE = Namespace.create("namespace");
+    private static final String TABLENAME = "test";
+    private static final TableReference DATA_TABLE_REF = TableReference.create(NAMESPACE, TABLENAME);
+
+    private static final int NUM_THREADS = 100;
+    private static final int NUM_ITERATIONS = 1_000;
+
+    private final AtomicLong timestamp = new AtomicLong();
+    private final KeyValueService rawKvs = new InMemoryKeyValueService(true);
+    private final KvTableMappingService tableMapper = KvTableMappingService.create(rawKvs, timestamp::incrementAndGet);
+    private final KeyValueService kvs = TableRemappingKeyValueService.create(rawKvs, tableMapper);
+
+    @Test
+    public void canConcurrentlyDropAndCreateDisjointSetsOfTables() {
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);
+        for (int i = 0; i < NUM_ITERATIONS; i++) {
+            String firstWorld = generateRandomTablesPrefixed();
+            String secondWorld = generateRandomTablesPrefixed();
+            final Set<Future<Void>> futures = new HashSet<>();
+            futures.add(executor.submit(() -> dropTablesWithPrefix(firstWorld)));
+            futures.add(executor.submit(() -> dropTablesWithPrefix(secondWorld)));
+            futures.forEach(Futures::getUnchecked);
+        }
+    }
+
+    @Test
+    public void canConcurrentlyReadFromStableTablesWhileCreatingAndDroppingOtherTables() {
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);
+        kvs.createTable(DATA_TABLE_REF, AtlasDbConstants.GENERIC_TABLE_METADATA);
+        Cell testCell = Cell.create(PtBytes.toBytes("row"), PtBytes.toBytes("col"));
+        byte[] value = PtBytes.toBytes("value");
+        long testTimestamp = 1L;
+        kvs.put(DATA_TABLE_REF, ImmutableMap.of(testCell, value), testTimestamp);
+        for (int i = 0; i < NUM_ITERATIONS; i++) {
```

Comment:
I have no idea what is going on here.

Reply:
Similar to the first test, except that instead of dropping table A and dropping table B concurrently, this drops table A and _reads_ from table B concurrently. They go through different code-paths IIRC (and both fail without this change)